src/myrtlespeech/post_process/transducer_decoder_base.py

```diff
@@ -0,0 +1,164 @@
+            audio_len = x[1][b].unsqueeze(0)
+            audio_features = x[0][b, :, :, :audio_len].unsqueeze(0)
+            audio_inp = (audio_features, audio_len)
+            sentence = self.decode(audio_inp)
+            preds.append(sentence)
+
+        # restore training state
+        self._model.train(training_state)
+
+        del audio_inp, audio_features, audio_len
+        return preds
+
+    def decode(self, inp: Tuple[torch.Tensor, torch.Tensor]) -> List[int]:
+        r"""Decodes a single sample.
+
+        Args:
+            inp: Tuple where the first element is the encoder
+                input (a :py:class:`torch.Tensor`) with size ``[batch,
+                channels, features, max_input_seq_len]`` and the second element
+                is a :py:class:`torch.Tensor` of size ``[batch]`` where each
+                entry represents the sequence length of the corresponding
+                *input* sequence to the rnn.  Note that ``inp`` is passed
+                straight to :py:class:`TransducerEncoder`.
+
+        Returns:
+            A List of length ``[batch]`` where each element is a List of
+            indexes.
+
+        """
+        raise NotImplementedError(
+            "decode method not implemented. Do not \
+            instantiate `TransducerDecoderBase` directly. Instead use a \
+            specific e.g. `TransducerGreedyDecoder`"
+        )
+
+    def _pred_step(self, label, hidden):
+        b"""Performs a step of the prediction net during inference."""
+        if label == self._SOS:
+            y = None
+        else:
+            if label > self._blank_index:
+                label -= 1  # Since input label indexes will be offset by +1
+                # for labels above blank. Avoiding this complexity is
+                # the reason for enforcing  blank_index = (len(alphabet) - 1)
+            y = collate_label_list([[label]], device=self._device)
+        (out, hid), lengths = self._model.predict_net.predict(
+            y, hidden, is_training=False
+        )
+        return (out, lengths), hid
+
+    def _joint_step(self, enc, pred):
+        r"""Performs a step of the model joint network during inference."""
+
+        logits, _ = self._model.joint((enc, pred))
+        res = torch.nn.functional.log_softmax(logits, dim=-1).squeeze()
+        assert (
+            len(res.shape) == 1
+        ), "this _joint_step result should have just one non-zero dimension"
+
+        return res
+
+    def _get_last_idx(self, labels):
+        b"""Returns the final index of a list of labels."""
+        return self._SOS if labels == [] else labels[-1]
+
+    def __repr__(self):
+        string = self._get_name() + "("
+        string += f"max_symbols_per_step={self._max_symbols_per_step}, "
+        string += f"blank_index={self._blank_index}"
+        if hasattr(self, "beam_width"):
+            string += f", beam_width={self.beam_width}"
+        string += ")"
+        return string
```

Comment:
Yes re: the parent classes not having knowledge of child classes, I have fixed this.

Reply:
I was actually using ``__repr__`` instead of `extra_repr`` intentionally in order to avoid the model definition appearing twice in the representation of a `seq_to_seq` object. When using `extra_repr` the printed `seq_to_seq` was as follows:

```
SpeechToText(
  (alphabet): Alphabet(symbols=[' ', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', "'", '_'])
  (model): Transducer(
    [... long model printout]
  )
  (loss): TransducerLoss(
    (transducer_loss): RNNTLoss()
  )
  (post_process): TransducerGreedyDecoder(
     (_model): Transducer(
        [... long model printout repeated]
  ),  max_symbols_per_step=100, blank_index=28)
)
```
which I personally found quite annoying. V. happy to return to this though - let me know if you would like me to.